### PR TITLE
First telemetry API draft

### LINKED ISSUE #169: add default, opt-out Google Analytics script for collecting generic usage information
We want usage information on approximate MAU and DAU to be able to gauge interest and usage of the [Unity WebVR Assets](https://u3d.as/1476) package [released earlier this month](https://hacks.mozilla.org/2018/02/create-vr-on-the-web-using-unity3d/).

Developers should be able to remove or configure this however they want. See #164 for documenting/messaging.

We have a Google Analytics 360 account set up already (ID: `UA-77033033-6`). It's injected by Unity on this project's [Asset Store detail page](https://u3d.as/1476).

@delapuente wrote a great docs page on [configuring Google Analytics](https://github.com/mozilla/unity-webvr-export/blob/master/docs/customization/adding-ga.md) for developers.

This page describes how to use GA's [`gtag.js`](https://github.com/mozilla/unity-webvr-export/blob/master/docs/customization/adding-ga.md#set-up) (scroll to the last few images of that section). Possibly we may want to use `analytics.js` instead.

We want to have a default system that does the following for users:

- respects the browser's Do-Not-Track feature (i.e., check `navigator.doNotTrack === '1'`)
- includes an option to **[opt out](https://developers.google.com/analytics/devguides/collection/analyticsjs/user-opt-out#opt-out_of_tracking_for_your_site)** of Mozilla's GA (i.e., add `window['ga-disable-UA-77033033-6'] = true`)

**Here's a [snippet of JavaScript](https://gist.github.com/cvan/78966124dd757c65e2f482b767052e44#file-ga-js) to possibly work from.**

Also, consider looking at [these GA plugins](https://github.com/MozillaReality/vr-mozilla-org/blob/master/static/js/ga.js) (which are used on https://vr.mozilla.org/).

<hr>

Parent issue: #165
Related issues:
- #170 for adding perf data,
- #167 for building the Help modal,
- #164 for documenting/messaging

### LINKED ISSUE #172: Create a new JavaScript template file `telemetry.js` for telemetry services
For handling and injecting Sentry for JS errors (see #166) and Google Analytics for usage + perf metrics (see #168 + #170).

<hr>

Parent issue: #165
Related issues:
- #168 for submitting a _Request for Data Collection Review_ (to Mozilla Data Metrics team)
- #169 for adding GA usage data,
- #170 for adding GA perf data,
- #166 for adding Sentry for catching JS errors,
- #167 for building the Help modal,
- #164 for documentation/messaging

### PR DESCRIPTION
**Pull request by delapuente**
Fix #172
Fix #169 

The PR adds some folder structure to the template and proposes an API for telemetry, Roughly, the telemetry module installs itself inside the `UnityLoader` object at `UnityLoader.WebVR.telemtry`. This way, there is only one entry point for everything.

To configure and start telemetry, the developer calls `telemetry.start(config)` and passes a `config` object with two properties so far granting permissions for Mozilla Research to log errors and gather analytics:

```js
MozillaResearch.telemetry.start({
  analytics: true,
  errorLogging: true        
});
```

If the developer wants to add their custom tracker, they should use `MozillaResearch.telemetry.ga.create()` with the same signature as the [`create` command](https://developers.google.com/analytics/devguides/collection/analyticsjs/command-queue-reference#create) of GA. The library then returns a function that can be treated like the regular `ga()` function. The pattern is inspired by @cvan [proposal](https://gist.github.com/cvan/78966124dd757c65e2f482b767052e44#file-ga-js) but provides more flexibility for the end user.

### CODE REVIEW

File: Assets/WebGLTemplates/WebVR/index.html [hunk 1]
```diff
@@ -10,8 +10,15 @@
     <title>%UNITY_CUSTOM_NAME% | %UNITY_WEB_NAME%</title>
     <meta name="description" content="%UNITY_CUSTOM_DESCRIPTION%">
     <link rel="icon" href="favicon.ico">
-    <link rel="stylesheet" type="text/css" href="webvr.css">
+    <link rel="stylesheet" type="text/css" href="styles/webvr.css">
     <script src="%UNITY_WEBGL_LOADER_URL%"></script>
+    <script src="src/telemetry.js"></script>
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/index.html]: this feels more like `lib/`

Comment (cvan) [reply Assets/WebGLTemplates/WebVR/index.html]: any Telemetry code should happen _before_ the Unity Engine is loaded. it's quite possible that there are errors that get thrown which could affect the loading of that and/or this script.

either way, the code should be run before the Unity JS. so we can get accurate perf timings, catch network errors, runtime errors, etc. either it should be a blocking `<script>` above or an inline `<script>`.

File: Assets/WebGLTemplates/WebVR/index.html [hunk 2]
```diff
@@ -76,8 +83,8 @@ <h3>You&rsquo;ll need a <a href="https://webvr.rocks/">WebVR-enabled browser</a>
       <button id="entervr" value="Enter VR"></button>
     </div>
 
-    <script src="gl-matrix-min.js"></script>
-    <script src="webvr-polyfill.min.js"></script>
+    <script src="lib/gl-matrix-min.js"></script>
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/index.html]: this should be called `vendor/`

File: Assets/WebGLTemplates/WebVR/src/telemetry.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
+/* global localStorage, location, Raven */
+(function (window) {
+'use strict';
+
+if (!('UnityLoader' in window)) {
+  console.warn('`UnityLoader` object not found.');
+  return;
+}
+
+var navigator = window.navigator;
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/src/telemetry.js]: this is unnecessary; just use `navigator` directly

File: Assets/WebGLTemplates/WebVR/src/telemetry.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
+/* global localStorage, location, Raven */
+(function (window) {
+'use strict';
+
+if (!('UnityLoader' in window)) {
+  console.warn('`UnityLoader` object not found.');
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/src/telemetry.js]: `not found` -> `not be found`

File: Assets/WebGLTemplates/WebVR/src/telemetry.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  }
+  if (config.researchAnalytics) {
+    startAnalytics();
+  }
+};
+
+setupAnalytics();
+
+function setupAnalytics() {
+  window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/src/telemetry.js]: I know this is copied from GA's snippet builder, but might as well just unminify this

File: Assets/WebGLTemplates/WebVR/src/telemetry.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  script.addEventListener('error', function (err) {
+    if (callback) {
+      callback(err);
+    }
+  });
+  document.head.appendChild(script);
+  return script;
+}
+
+function getModule (modulePath) {
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/src/telemetry.js]: maybe I'm missing the purpose: why we're using this as opposed to just calling the object directly?

File: Assets/WebGLTemplates/WebVR/index.html [hunk 7]
```diff
@@ -10,8 +10,15 @@
     <title>%UNITY_CUSTOM_NAME% | %UNITY_WEB_NAME%</title>
     <meta name="description" content="%UNITY_CUSTOM_DESCRIPTION%">
     <link rel="icon" href="favicon.ico">
-    <link rel="stylesheet" type="text/css" href="webvr.css">
+    <link rel="stylesheet" type="text/css" href="styles/webvr.css">
     <script src="%UNITY_WEBGL_LOADER_URL%"></script>
+    <script src="src/telemetry.js"></script>
+    <script>
+      UnityLoader.WebVR.telemetry.start({
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/index.html]: I like the invocation and naming of this. good work! :+1:

although I'm not keen on this being attached to the `UnityLoader` for the aforementioned reasons.

File: Assets/WebGLTemplates/WebVR/src/telemetry.js (hunk 3 above)
Comment (delapuente) [reply Assets/WebGLTemplates/WebVR/src/telemetry.js]: I we plan to test in a future, this is more convenient, don't you think?

File: Assets/WebGLTemplates/WebVR/lib/telemetry.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
+/* global localStorage, location, Raven */
+(function (window) {
+'use strict';
+
+if (!('MozillaResearch' in window)) {
+  window.MozillaResearch = {};
+}
+
+if (!('telemetry' in window.MozillaResearch)) {
+  window.MozillaResearch.telemetry = {} ;
```
Comment (cvan) [inline Assets/WebGLTemplates/WebVR/lib/telemetry.js]: trailing whitespace

**Comment by delapuente** [comment]
Makes sense. I'm going to undo the changes to CSS, remove the trailing whitespace and get it merged.

**Pull request merged**